cpp/src_prims/selection/knn.h

```diff
@@ -104,93 +170,451 @@ void merge_tables(long n, long k, long nshard, float *distances, long *labels,
+    // for offsets of the indices, build them
+    // from the local partitions
+    id_ranges = new std::vector<int64_t>();
+    int64_t total_n = 0;
+    for (int i = 0; i < n_params; i++) {
+      if (i < input.size()) {
+        id_ranges->push_back(total_n);
+      }
+      total_n += sizes[i];
+    }
+  } else {
+    // otherwise, use the given translations
+    id_ranges = translations;
+  }
 
-  IntType total_n = 0;
+  int device;
+  CUDA_CHECK(cudaGetDevice(&device));
 
-  for (int i = 0; i < n_params; i++) {
-    if (i < n_params)  // if i < sizes[i]
-      id_ranges->push_back(total_n);
-    total_n += sizes[i];
-  }
+  device_buffer<int64_t> trans(allocator, userStream, id_ranges->size());
+  updateDevice(trans.data(), id_ranges->data(), id_ranges->size(), userStream);
 
-  float *result_D = new float[k * size_t(n)];
-  long *result_I = new long[k * size_t(n)];
+  device_buffer<float> all_D(allocator, userStream, input.size() * k * n);
+  device_buffer<int64_t> all_I(allocator, userStream, input.size() * k * n);
 
-  float *all_D = new float[n_params * k * size_t(n)];
-  long *all_I = new long[n_params * k * size_t(n)];
+  if (n_int_streams > 0) CUDA_CHECK(cudaStreamSynchronize(userStream));
 
-  ASSERT_DEVICE_MEM(search_items, "search items");
-  ASSERT_DEVICE_MEM(res_I, "output index array");
-  ASSERT_DEVICE_MEM(res_D, "output distance array");
+  for (int i = 0; i < input.size(); i++) {
+    const float *ptr = input[i];
+    IntType size = sizes[i];
 
-  CUDA_CHECK(cudaStreamSynchronize(s));
+    try {
+      faiss::gpu::StandardGpuResources gpu_res;
 
-#pragma omp parallel
-  {
-#pragma omp for
-    for (int i = 0; i < n_params; i++) {
-      const float *ptr = input[i];
-      IntType size = sizes[i];
+      cudaStream_t stream =
+        n_int_streams > 0 ? internalStreams[i % n_int_streams] : userStream;
 
-      cudaPointerAttributes att;
-      cudaError_t err = cudaPointerGetAttributes(&att, ptr);
+      gpu_res.noTempMemory();
+      gpu_res.setCudaMallocWarning(false);
+      gpu_res.setDefaultStream(device, stream);
 
-      if (err == 0 && att.device > -1) {
-        CUDA_CHECK(cudaSetDevice(att.device));
-        CUDA_CHECK(cudaPeekAtLastError());
+      faiss::gpu::bruteForceKnn(&gpu_res, faiss::METRIC_L2, ptr, rowMajorIndex,
+                                size, search_items, rowMajorQuery, n, D, k,
+                                all_D.data() + (i * k * n),
+                                all_I.data() + (i * k * n));
 
-        try {
-          faiss::gpu::StandardGpuResources gpu_res;
+      CUDA_CHECK(cudaPeekAtLastError());
 
-          cudaStream_t stream;
-          CUDA_CHECK(cudaStreamCreate(&stream));
+    } catch (const std::exception &e) {
+      std::cout << "Exception occurred: " << e.what() << std::endl;
+    }
+  }
 
-          gpu_res.noTempMemory();
-          gpu_res.setCudaMallocWarning(false);
-          gpu_res.setDefaultStream(att.device, stream);
+  for (int i = 0; i < n_int_streams; i++) {
```

Comment:
and when `n_int_streams` is 0, we'll need to sync on `userStream` instead, no?

Reply:
I don't think so because we're no longer copying to host. Any further kernels queued up to the `userStream` should be scheduled in serial order so we leave it up to the caller to synchronize.